### PR TITLE
Support metaclass subclasses

### PR DESCRIPTION
**Pull request by mbustosorg**
Descendants should check if the class is a Metaclass.  Added test also.

### CODE REVIEW

File: pdoc/__init__.py [hunk 1]
```diff
@@ -1210,6 +1210,10 @@ def subclasses(self) -> List['Class']:
         The objects in the list are of type `pdoc.Class` if available,
         and `pdoc.External` otherwise.
         """
+        if issubclass(self.obj, type):
+            return [self.module.find_class(c)
+                    for c in self.obj.__subclasses__(self.obj)]
+
```
Comment (kernc) [inline pdoc/__init__.py]: I think the whole method can be replaced with:
```py
return [self.module.find_class(c)
        for c in type.__subclasses__(self.obj)]
```
This makes the branch redundant as it works for both cases.

**Comment by mbustosorg** [comment]
d.py is tested as part of the example_pkg test.

Calling with self.obj doesn't work because non metaclass type classes take
no arguments.  The tests fail that way.

Thanks,
mauricio

**Comment by kernc** [comment]
I think that's because `self.obj.__subclasses__` is a bound method whereas `type.__subclasses__` is unbound and can be passed the `self.obj` (meta)class parameter. Seems to work.

**Comment by mbustosorg** [comment]
I tried that iteration and it failed when testing d.py.  Did you see different results?

**Comment by kernc** [comment]
Seems to work here:
```py
>>> class A: pass
>>> class B(type): pass
>>> class C(A): pass
>>> class D(B): pass
>>> type.__subclasses__(A)
[<class '__main__.C'>]
>>> type.__subclasses__(B)
[<class '__main__.D'>]
```

In addition, would you consider creating a new test `ApiTest.test_subclasses()` where you create a few local classes and call `pdoc.Class(cls.__name__, sys, cls).subclasses()` on them?

**Comment by mbustosorg** [comment]
Yes.  I'll have a look at why I'm seeing different results as you and will
create the new test case.

File: pdoc/test/__init__.py [hunk 2]
```diff
@@ -467,6 +467,23 @@ def test_inherited_members(self):
...
+        class B(type):
+            pass
+
+        class C(A):
+            pass
+        
+        class D(B):
+            pass
+        a_subclasses = type.__subclasses__(A)
+        b_subclasses = type.__subclasses__(B)
```
Comment (kernc) [inline pdoc/test/__init__.py]: Let's instead test pdoc's API (`pdoc.Class.subclasses`):
```py
any_module = pdoc.Module(pdoc)
self.assertEqual(pdoc.Class('A', any_module, A).subclasses(), [C])
```

File: pdoc/test/example_pkg/_test_linking/d.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class D(type):
+
+    def __init__(cls, name, bases, dct):
+        super().__init__(cls, name, bases, dct)
+        cls._instance = None
+
+    def __call__(cls, *args, **kwargs):
+        if cls._instance is None:
+            cls._instance.__call__(*args, **kwargs)
+        return cls._instance
```
Comment (kernc) [inline pdoc/test/example_pkg/_test_linking/d.py]: Afterwards, this file is obsolete.

**Comment by mbustosorg** [comment]
So you would recommend creating a new submodule that includes these classes
instead of in the code like this?

**Comment by kernc** [comment]
No, no, I love how few extra lines are needed to define those classes! Just test _pdoc_ API (`pdoc.Class.subclasses`) instead of Python API (`type` dunder methods). :smile:

**Comment by kernc** [comment]
Sorry, with `any_module` I meant it can be a dummy module just so long the constructor executes.

**Comment by kernc** [comment]
Applied with slight modification. I see I have been oversimplifying my examples. Thanks for figuring it out! :smile:

**Pull request merged**